### ISSUE TITLE
JavaScript Data Types

### ISSUE THREAD

**Issue by Zakintaliban** [belajar-3-bahasa/the-course#3]
# JavaScript Data Types

Most of the time, a JavaScript application needs to work with information. Here are two examples:

- An online shop – the information might include goods being sold and a shopping cart.
- A chat application – the information might include users, messages, and much more.

Variables are used to store this information.

There are 7 basic types in JavaScript.

- `number` for numbers of any kind: integer or floating-point.
- `string` for strings. A string may have one or more characters, there’s no separate single-character type.
- `boolean` for `true`/`false`.
- `null` for unknown values – a standalone type that has single value null.
- `undefined` for unassigned values – a standalone type that has a single value undefined.
- `object` for more complex data structures.
- `symbol` for unique identifiers.

The `typeof` operator allows us to see which type is stored in a variable.

- Two forms: `typeof x` or `typeof(x)`.
- Returns a string with the name of the type, like `"string"`.
- For `null` returns `"object"` – this is an error in the language, it’s not an object.

Remember, blob assets such as images, photos, videos are not JavaScript data types. So actually we cannot store those inside the JavaScript program. But we can access it through the file system.

---

## Dynamic Typing

A variable in JavaScript can contain any data. A variable can at one moment be a string and at another be a number:

Programming languages that allow such things are called “dynamically typed”, meaning that there are data types, but variables are not bound to any of them.

```js
let message = "hello";
message = 123456;
```

---

## Number

The number type represents both integer and floating-point numbers.

```js
let n = 123;
n = 12.345;
```

```js
let integerCalculation = 1 + 2; // gives 3

let weirdCalculation = 0.1 + 0.2; // 0.3 expected, actual answer 0.30000000000000004
```

---

## String

A string in JavaScript must be surrounded by quotes.

```js
let text = "Hello World";
let textOther = "Single quotes are okay too";
let textLonger = `We can embed ${text}`;
```

```js
let firstName = "Will";
let lastName = "Alexander";
```

```js
let wholeName = firstName + " " + lastName; // value: "Will Alexander"
```

In JavaScript, there are 3 types of quotes.

- Double quotes:

```js
"Hello";
```

- Single quotes:

```js
"Hello";
```

- Backticks:

```js
`Hello`;
```

Double and single quotes are “simple” quotes. There’s no difference between them in JavaScript.

Backticks are “extended functionality” quotes. They allow us to embed variables and expressions into a string by wrapping them in `${…}`, for example:

```js
let name = "John";

// embed a variable
console.log(`Hello, ${name}!`); // Hello, John!

// embed an expression
console.log(`the result is ${1 + 2}`); // the result is 3

// multiple lines
console.log(`Title
Line One
Line Two
`);

// multiple lines affected by a line break and indentation
console.log(`
  # Title
  
  - Line One
  - Line Two
`);
```

In some languages, there is a special “character” type for a single character. For example, in the C language and Java, it is `char`.

In JavaScript, there is no such type. There’s only one type: `string`. A string may consist of only one character or many of them.

---

## Boolean (logical type)

The boolean type has only two values: `true` and `false`.

```js
let userIsSignedIn = true;
let userIsAdmin = false;
```

This type is commonly used to store yes/no values: `true` means “yes, correct”, and `false` means “no, incorrect”.

```js
let nameFieldChecked = true; // yes, name field is checked
let ageFieldChecked = false; // no, age field is not checked
```

Boolean values also come as a result of comparisons:

```js
let isGreater = 4 > 1;

alert(isGreater); // true (the comparison result is "yes")
```

---

## `null`

The special null value does not belong to any of the types described above.

It forms a separate type of its own which contains only the null value:

```js
let age = null;
```

In JavaScript, null is not a “reference to a non-existing object” or a “null pointer” like in some other languages.

It’s just a special value that represents “nothing”, “empty” or “value unknown”.

`null` is not the same as `undefined`.

---

## `undefined`

The special value undefined also stands apart. It makes a type of its own, just like null.

The meaning of undefined is “value is not assigned”.

If a variable is declared, but not assigned, then its value is undefined:

```js
let x;

alert(x); // shows "undefined"
```

Technically, it is possible to assign undefined to any variable:

```js
let x = 123;

x = undefined;

alert(x); // "undefined"
```

---

## Arrays

To create an empty array and store it in a variable, you use a pair of square brackets:

```js
let guests = [];
```

You can also create a populated array by putting the desired elements inside those square brackets:

```js
let guests = ["Sarah Kate", "Audrey Simon", "Will Alexander"];
```

You can then access elements of that array by their index:

```js
let firstGuest = guests[0]; // 'Sarah Kate'
let thirdGuest = guests[2]; // 'Will Alexander'
let undefinedGuest = guests[12]; // undefined
```

A very important side note - value vs. reference

In JavaScript, primitive types like numbers, booleans, and strings are passed by value. This means that when you do something like this:

```js
let numberOfGuests = 20;

let totalNumberOfGuests = numberOfGuests; // 20
```

It is the value 20, which is copied into the new variable, and no link is maintained between the two variables.

This is not the case with objects and arrays, which are passed by reference. If you are not aware of this, it can lead to some unexpected behaviors. For example:

```js
let artistProfile = {
  name: "Tau Perkington",
  age: 27,
  available: true
};

let allProfiles = [artistProfile]; // new Array containing the above object

artistProfile.available = false; // changing the object

console.log(allProfiles); // will show { name: 'Tau Perkington', age: 27, available: false }
```

Even though we created the array and passed it the object before modifying the object, you still see the modification in the array. This is because when using arrays and objects, you are passing references to the objects, as opposed to the value of the data within them. The variables artistProfile and allProfiles shown above contain references to the object and array in memory.

This can seem a bit tricky to understand, but with practice, you'll get the hang of it!

Counting elements
The length property of an array tells you how many elements there are in it:

```js
let guests = ["Will Alexander", "Sarah Kate", "Audrey Simon"];
let howManyGuests = guests.length; // 3
```

Use dot notation to access the length property of your array. This may not be very useful in this example (seeing as we know how many guests we put in the array!), but there are many situations in which you will not know how many elements an array has in advance.

Adding and removing elements
To add an element to the end of an array, use its push method:

```js
guests.push("Tau Perkington"); // adds 'Tau Perkington' to the end of our guests array
```

You use dot notation to access the array's push method and put the element you wish to add in parentheses.

If you want to add your element at the beginning of the array instead of the end, use the unshift method:

```js
guests.unshift("Tau Perkington"); // 'Tau Perkington' is added at the beginning of the guests array
```

To remove the last element of an array, call its pop method, without passing any arguments:

```js
guests.pop(); // removes the last element from the array
```

---

## Objects

All other types are called “primitive” because their values can contain only a single thing (be it a string or a number or whatever). In contrast, objects are used to store collections of data and more complex entities.

```js
const human = {
  name: "Brendan Eich",
  website: "https://brendaneich.com",
  twitter: "https://twitter.com/BrendanEich",
  github: "https://github.com/BrendanEich",
  "about-me":
    "Brendan Eich is an American technologist and creator of the JavaScript programming language"
  // multiword property name must be quoted
};
```

```js
let myBook = {
  title: "The Story of Tau",
  author: "Will Alexander",
  numberOfPages: 250,
  isAvailable: true
};
```

```js
let myBook = {
  title: "The Story of Tau",
  author: "Will Alexander",
  numberOfPages: 250,
  isAvailable: true
};

let bookTitle = myBook.title; // 'The Story of Tau'
let bookPages = myBook.numberOfPages; // 250
```

---

## Classes

Building an object by hand using curly-brace notation is fine for simple, single objects. However, you will often need many objects of the same type. This is where classes come in handy.

As mentioned earlier, a class is a blueprint for an object in code. It allows you to build multiple objects of the same type (called instances of the same class) more quickly, easily, and reliably.

Let's have a look at how you build a class in code:

To create a class in JavaScript, use the class keyword, followed by a name. Then enclose the class code in curly braces:

```js
class Book {}
```

For this class, we would like every Book to have a title, an author, and several pages. To do this, you use what is called a constructor:

```js
class Book {
  constructor(title, author, pages) {}
}
```

There is a set of instructions to follow within the constructor when you create a new instance of the Book class. To assign the title, author and number of pages you receive to this instance, use this keyword and dot notation:

```js
class Book {
  constructor(title, author, pages) {
    this.title = title;
    this.author = author;
    this.pages = pages;
  }
}
```

Here, this keyword refers to the new instance. Therefore, it is using dot notation to assign the received values to its corresponding keys.

Now that the class is complete, you can create new instances of it using the new keyword:

```js
let myBook = new Book("The Story of Tau", "Will Alexander", 250);
```

This line creates the following object:

```js
{
    title: 'The Story of Tau',
    author: 'Will Alexander',
    pages: 250
}
```

Having a Book class allows you to create new Book objects quickly and easily.

Right! Now it's your turn. The best way to learn about classes is to make one yourself.

---

## `typeof` Operator

The typeof operator returns the type of the argument. It’s useful when we want to process values of different types differently or just want to do a quick check.

It supports two forms of syntax:

- As an operator: `typeof x`.
- As a function: `typeof(x)`.

In other words, it works with parentheses or without them. The result is the same.

The call to typeof x returns a string with the type name:

```js
typeof "foo"; // "string"
typeof 0; // "number"
typeof true; // "boolean"
typeof undefined; // "undefined"
```

---
